text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,144 @@
+- `bool obs_enum_output_service_types(const char *protocol, size_t idx, const char **id)`: enumerate all outputs types compatible with the given protocol.
+- `bool obs_enum_output_protocols_prefixes(size_t idx, const char **prefix)`: enumerate all registered URL prefixes.
+- `bool obs_enum_output_protocols(size_t idx, const char **prefix)`: enumerate all registered protocol.
+- `const char *obs_get_output_supported_video_codecs(const char *id)`: return compatible video codecs of the given output id.
+- `const char *obs_get_output_supported_audio_codecs(const char *id)`: return compatible audio codecs of the given output id.
+
+### About protocols
+
+- HLS relies on HTTP/HTTPS and it is not the only existing protocol in this case. This protocol has no unique URL prefixes.
+
+- Because of the deprecation of the protocol, FTL will not have a URL prefix asign to its output to not make it available through custom server in the UI.
+
+## Services API
+Since a streaming service may not accept all the codecs usable by a protocol, adding a way to set supported codecs is required.
+
+Adding to `obs_service_info`:
+- `const char *(*get_protocol)(void *data)`: return the protocol used by the service. RFC 39 will allow multi-protocol services in the future.
+- `const char **(*get_supported_video_codecs)(void *data)`: video codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+- `const char **(*get_supported_audio_codecs)(void *data)`: audio codecs supported by the service. Optional, fallback to protocol supported codecs if not set.
+
+Adding to this API these functions:
+- `const char *obs_service_get_protocol(const obs_service_t *service)`: return the protocol used by the service.
+- `const char **obs_service_get_supported_video_codecs(const obs_service_t *service)`: return video codecs compatible with the service.
+- `const char **obs_service_get_supported_audio_codecs(const obs_service_t *service)`: return audio codecs compatible with the service.
+
+### Services and connection informations
+
+Depending on the protocol, the service object should provide various types of connection details (e.g. server URL, stream key, username, password).
+
+Currently, `get_key` can provide a stream id (SRT) or an encryption passphrase (RIST) rather than a stream key.
+
+Rather than adding a getter for each possible type of information, a getter where we can choose which type of information we want should be implemented.
+
+Each type of connection details will be defined by a macro with an even number value. Odd number values will be reserved for potential third-party protocols (RFC 39).
+
+List of types:
+```c
+#define OBS_SERVICE_INFO_SERVER_URL 0
+#define OBS_SERVICE_INFO_STREAM_KEY 2
+#define OBS_SERVICE_INFO_USERNAME 4 
+#define OBS_SERVICE_INFO_PASSWORD 6
+#define OBS_SERVICE_INFO_STREAM_ID 8
+#define OBS_SERVICE_INFO_ENCRYPT_PASSPHRASE 10
```

Comment:
Can we get `#define OBS_SERVICE_INFO_BEARER_TOKEN 12` for WebRTC?

Reply:
Currently, OBS Studio only supports RTMP(S), FTL (deprecated), HLS, SRT, RIST.